reference/promise-types/custom.markdown

```diff
@@ -95,6 +96,13 @@ Due to the implementation details, the following attributes from the `classes` b
 * `persist_time`
 * `timer_policy`
 
+### Evaluation passes and normal order
+
+In CFEngine, each bundle is evaluated in multiple passes (3 main passes for most promise types).
+Within each evaluation pass of a bundle, the promises are not evaluated from top to bottom, but based on a [normal order][Normal Ordering] of the promise types.
+Custom promise types are added dynamically and don't have a predefined order, they are evaluated as they appear within a bundle (top to bottom), but at the end of each evaluation pass, after all the built in promise types.
```

Comment:
Evaluated as they appear?

So, this policy file http promise would happen after reports and before git:

```
http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));
  git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";
```

And in another policy file, git would happen after reports before http:

```
git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";
http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));
```

Does that also apply to those that are interleaved? Meaning that after reports the git promise for starter pack would get actuated, and then the http promise type and then the git promise type for documentation? 

```
git:
    "starter_pack_repo"
      repository => "https://github.com/cfengine/starter_pack",
      destination => "/northern.tech/cfengine/starter-pack",
      version => "master";

http:
    "https://cfengine.com/images/cfengine-logo.svg"
      file => "/var/cfengine/cfengine-logo.svg",
      if => not(fileexists("/var/cfengine/cfengine-logo.svg"));

git:
    "cfengine-documentation"
      repository => "https://github.com/cfengine/documentation",
      destination => "/northern.tech/cfengine/documentation",
      version => "master";
```

Reply:
The _promise types_ are evaluated in the order they appear, not the individual promises.

Just like with built in promise types, adding repeating sections of the same promise types or even interleaving them has no effect, it's the same as putting all the promises with the same type in one section.

I think we should change it, it makes more sense to me if all the promises were just evaluated in order, but at least it's somewhat consistent now.

Reply:
OK, and it's about their USE within a bundle, **not** the order in which they are defined (order of promise agent blocks)